models, sundog, prairiedog: add kernel boot config support

etungsten opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
N/A


**Description of changes:**
```
    models: new boot settings to specify kernel boot config

    This adds `settings.boot` to help configure kernel boot config settings.
```
```
    prairiedog: add boot config functionality

     - Adds new subcommand `generate-boot-config` to generate an initrd
    image file with boot config contents populated from Bottlerocket boot
    settings.
     - Adds new subcommand `generate-boot-settings` to generate bottlerocket
    settings from kernel's bootconfig file at `/proc/bootconfig`

    Moves prairiedog from `sources/prairiedog` to `sources/api/prairiedog`
    now that it needs to query from the Bottlerocket API and need settings
    model information.
```
```
    sundog: update settings retrieval to use prefix

    This updates the settings retrieval to use prefixes for matching and
    querying. We cannot use full keys for matching because we now have
    settings generator for keys that map to top level structs that
    encompasses multiple settings.

```
```
    models: add metadata for 'settings.boot'

    Add settings generator for `settings.boot` and restart commands for when
    boot settings change.
```
```
    migrations: add migrations for new `settings.boot` setting
```
```
    README: add description and example for 'settings.boot'
```
```
    metal: remove kernel console settings from variant metadata

    We've added bootconfig support so console settings should be specified
    through there. The console settings specified to the kernel commandline
    will always take effect, so we're removing them from the variant
    metadata.

```

**Testing done:**
Launched instance with the following user-data:
```
[settings.boot]
kernel.test1 = ["bark", "barkbark"]
kernel.test2 = ["woof", "woofwoof"]
init.test1 = ["good", "morning"]
init.test2 = ["hello", "goodbye"]
```
Prairiedog successfully generates the initrd image file containing boot config:
```
[ec2-user@ip-192-168-10-10 ~]$ hexdump -C bootconfig.data
00000000  69 6e 69 74 2e 74 65 73  74 31 20 3d 20 22 67 6f  |init.test1 = "go|
00000010  6f 64 22 2c 20 22 6d 6f  72 6e 69 6e 67 22 0a 69  |od", "morning".i|
00000020  6e 69 74 2e 74 65 73 74  32 20 3d 20 22 68 65 6c  |nit.test2 = "hel|
00000030  6c 6f 22 2c 20 22 67 6f  6f 64 62 79 65 22 0a 6b  |lo", "goodbye".k|
00000040  65 72 6e 65 6c 2e 74 65  73 74 31 20 3d 20 22 62  |ernel.test1 = "b|
00000050  61 72 6b 22 2c 20 22 62  61 72 6b 62 61 72 6b 22  |ark", "barkbark"|
00000060  0a 6b 65 72 6e 65 6c 2e  74 65 73 74 32 20 3d 20  |.kernel.test2 = |
00000070  22 77 6f 6f 66 22 2c 20  22 77 6f 6f 66 77 6f 6f  |"woof", "woofwoo|
00000080  66 22 00 00 84 00 00 00  fb 29 00 00 23 42 4f 4f  |f".......)..#BOO|
00000090  54 43 4f 4e 46 49 47 0a                           |TCONFIG.|
00000098
```
Verified that it's a valid initrd image file with valid boot config content with the linux kernel's `bootconfig` tool:
```
$ linux/tools/bootconfig/bootconfig -l bootconfig.data
init.test1 = "good", "morning"
init.test2 = "hello", "goodbye"
kernel.test1 = "bark", "barkbark"
kernel.test2 = "woof", "woofwoof"
```

- [x] Integrated test on real bare-metal hosts with a provisioned boot config file

Built a `metal-dev` image and `dd`'d it to a baremetal server's `/dev/sda`.
I manually provisioned a bootconfig.data file that contained the following bootconfig:
```
$ ./bootconfig -l bootconfig.data
kernel.console = "tty0", "ttyS1,115200n8"
init.systemd.log_level = "debug"
```
and copied it into the `BOTTLEROCKET-PRIVATE` partition.
Then I booted the baremetal host from its harddrive.

Connected to the server's serial port over IP and saw that the host comes up successfully.
Checked settings and it's generated correctly from `/proc/bootconfig`:
```
[ec2-user@admin]$ apiclient get settings.boot
{
  "settings": {
    "boot": {
      "init": {
        "systemd.log_level": [
          "debug"
        ]
      },
      "kernel": {
        "console": [
          "tty0",
          "ttyS1,115200n8"
        ]
      }
    }
  }
}
```
If I check systemd, the log-level is correctly set to "debug"
```
bash-5.0# systemd-analyze get-log-level
debug
```

If I change the settings, `/var/lib/bottlerocket/bootconfig.data` gets updated as expected:
```
[ec2-user@admin]$ apiclient set --json '{"boot":{"kernel":{"console":["tty0","ttyS1,115200n8"]},"init":{"systemd.log_level":["info"]}}}'
[ec2-user@admin]$ cat /.bottlerocket/rootfs/var/lib/bottlerocket/bootconfig.data
init.systemd.log_level = "info"
kernel.console = "tty0", "ttyS1,115200n8"L#BOOTCONFIG
```

If I reboot, I can see that the systemd log-level gets updated to `info` as expected.
```                                                                                                                  
bash-5.0# cat /proc/bootconfig                                                                                                     
init.systemd.log_level = "info"                                                                                                    
kernel.console = "tty0", "ttyS1,115200n8"
                                                                              
bash-5.0# systemd-analyze get-log-level                                                                                            
info
```

- [x] Integrated test on real bare-metal hosts with a provisioned user-data.toml file

Built a `metal-dev` image and `dd`'d it to a baremetal server's `/dev/sda`.
I manually provisioned a user-data.toml file that looks like this:
```
bash-5.0# cat /var/lib/bottlerocket/user-data.toml 
[settings.boot.kernel]
"console" = ["tty0", "ttyS1,115200n8"]
[settings.boot.init]
"systemd.log_level" = ["debug"]
```
and copied it into the `BOTTLEROCKET-PRIVATE` partition.
Then I booted the baremetal host from its harddrive and everything comes up fine.

The generated bootconfig.data file is as expected:
```
[ec2-user@admin]$ hexdump -C /.bottlerocket/rootfs/var/lib/bottlerocket/bootconfig.data
00000000  69 6e 69 74 2e 73 79 73  74 65 6d 64 2e 6c 6f 67  |init.systemd.log|
00000010  5f 6c 65 76 65 6c 20 3d  20 22 64 65 62 75 67 22  |_level = "debug"|
00000020  0a 6b 65 72 6e 65 6c 2e  63 6f 6e 73 6f 6c 65 20  |.kernel.console |
00000030  3d 20 22 74 74 79 30 22  2c 20 22 74 74 79 53 31  |= "tty0", "ttyS1|
00000040  2c 31 31 35 32 30 30 6e  38 22 00 00 4c 00 00 00  |,115200n8"..L...|
00000050  08 18 00 00 23 42 4f 4f  54 43 4f 4e 46 49 47 0a  |....#BOOTCONFIG.|
00000060
```

If I reboot the host, then the bootconfig above gets loaded by the kernel as expected.
```
bash-5.0# cat /proc/cmdline                                                                                                        
console="tty0" console="ttyS1,115200n8" .....
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

etungsten committed:
models: new boot settings to specify kernel boot config

This adds `settings.boot` to help configure kernel boot config settings.

etungsten committed:
models: add metadata for 'settings.boot'

Add settings generator for `settings.boot` and restart commands for when
boot settings change.

etungsten committed:
migrations: add migrations for new `settings.boot` setting

etungsten committed:
README: add description and example for 'settings.boot'

webern commented on Release.toml:
```diff
@@ -112,3 +112,8 @@ version = "1.7.0"
     "migrate_v1.7.0_public-admin-container-v0-8-0.lz4",
     "migrate_v1.7.0_public-control-container-v0-6-0.lz4",
 ]
+"(1.7.0, 1.7.1)" = []
```
It may be a good idea to have a placeholder for a point release, but I don't think this has been our practice. Are there any ramifications for automated upgrade downgrade testing?

etungsten replied:
I'm not sure if I understand the question. Are you saying to put something like `"(1.7.0, 1.7.x)" = []` instead? Normally we'd just update the PR with the migrations as it becomes clear which version this feature will land in.

webern commented on packages/os/Cargo.toml:
```diff
@@ -13,7 +13,6 @@ source-groups = [
     "metricdog",
     "parse-datetime",
     "ghostdog",
-    "prairiedog",
```
Was prairiedog's inclusion here incorrect? I'm not sure what this does.

etungsten replied:
I moved prairiedog from `sources/prairiedog` to `sources/api/prairiedog` now that it needs to query from the Bottlerocket API and need settings model information. And `api` is its own buildsys "source-group".

webern commented on sources/api/migration/migrations/v1.8.0/boot-setting-metadata/src/main.rs:
```diff
@@ -0,0 +1,23 @@
+
+/// We added a new setting and generator for kernel boot configuration
+fn run() -> Result<()> {
+    migrate(AddMetadataMigration(&[SettingMetadata {
```
Is it possible to combine these two migrations into one? I think there may be a reason we cannot, but I forgot why.

etungsten replied:
It is not possible unless we introduce a new migration helper that does both within the same forward/backward migration. Each migration can only have a single `migrate` invocation. Because we only write out the datastore once. Any subsequent calls within the same migration binary invocation will just overwrite what was migrated previously.

webern commented on sources/models/src/aws-ecs-1/mod.rs:
```diff
@@ -20,6 +20,7 @@ struct Settings {
     ntp: NtpSettings,
     network: NetworkSettings,
     kernel: KernelSettings,
+    boot: BootSettings,
```
Do we actually want to add this to all variants? Or should we only add it where we know it is needed (metal)?

etungsten replied:
We do eventually want this for all variants and there's no reason to exclude non-metal variants from being able to modify boot kernel parameters.

bcressey replied:
In #1980 we discovered that we need to limit this to newer variants where we know for sure that all production releases have the necessary GRUB configuration.

webern commented on sources/api/prairiedog/src/de.rs:
```diff
@@ -0,0 +1,593 @@
+                        && c != ','
+                        && c != ';'
+                        && c != '#'
+                        && c != '}'))
```
This hurts my brain. Maybe a tiny helper function named `requires_quotes` so the conditional becomes `quoted || !requires_quotes`.

webern commented on sources/api/prairiedog/src/de.rs:
```diff
@@ -0,0 +1,593 @@
+        let mut is_quoted = false;
+        let mut start_index = 0;
+        for (i, c) in self.input.chars().enumerate() {
+            if c == '\'' || c == '\"' {
```
I guess we allow this `"starts with double quotes ends with single quote'`, but that seems fine.

etungsten replied:
Nice catch. I should actually make sure the opening and closing quote-types match.

etungsten replied:
Actually, `deserialize_str` does ensure that the values start and end with the same quote type, and we dispatch to deserialize a string value.

In this case of parsing out individual values from a sequence, we still need to ensure the quote-types match to make sure whether a `,` is within quotes or not.

webern commented on sources/api/prairiedog/src/de.rs:
```diff
@@ -0,0 +1,593 @@
+            } else {
+                // We've encountered the delimiter, and if it's outside quotes, we have a new element
+                if c == ',' && !is_quoted {
+                    elements.push(&self.input[start_index..i]);
```
Does it matter if each element includes vs. does-not-include the quotes in the string?

etungsten replied:
Each boot config value only contains printable characters or spaces except for delimiters such as semicolon, newline, comma, hash, and closing brace. These delimiters are only usable if the value itself is quoted with single-quotes or double-quotes.

So for a list of values delimited by commas, we need to make sure we keep track of quotes so if a `,` shows up in a value that is double quoted, we treat that as a singular value.

etungsten replied:
Ah, I misunderstood the question. The right behavior should be that single-quotes are allowed if the value is quoted with double-quotes and vice versa.

webern commented on sources/api/prairiedog/src/ser.rs:
```diff
@@ -0,0 +1,537 @@
+    }
+}
+
+/// Most types are invalid so we need to return an error in most cases.  This simplifies the creation of that error, with a customizable message for the type.
```
Wrap me

webern commented on sources/api/prairiedog/src/ser.rs:
```diff
@@ -0,0 +1,537 @@
+        self.output
+            .push_str(&self.prefix.context(error::BareStringSnafu)?.join("."));
+        self.output.push_str(" = ");
+        self.output.push_str(&format!("\"{}\"", v));
```
We don't preserve original quoting, but that seems fine.

webern commented on sources/api/prairiedog/src/error.rs:
```diff
@@ -0,0 +1,68 @@
+    #[snafu(display("Invalid boot config file, expected key-value entries for each line"))]
+    InvalidBootConfig,
+
+    #[snafu(display("Invalid boot config key encountered '{}'. Boot config keywords may only contain alphabets, numbers, dashes or underscores", key))]
```
```suggestion
    #[snafu(display("Invalid boot config key encountered '{}'. Boot config keywords may only contain ASCII alphas, numbers, dashes or underscores", key))]
```

webern commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +286,109 @@ fn load_crash_kernel() -> Result<()> {
 }
 
+/// Queries Bottlerocket boot settings and generates initrd image file with boot config as the only data
+async fn generate_boot_config(socket_path: String) -> Result<()> {
```
Is `String` actually required instead of `&str`?

webern commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +286,109 @@ fn load_crash_kernel() -> Result<()> {
+    // Calculate checksum of the boot config file
+    let mut checksum: u32 = 0;
+    for byte in bootconfig {
+        checksum += u32::from(*byte);
```
Maybe [checked add](https://doc.rust-lang.org/std/primitive.u32.html#method.checked_add) to be extra safe?

webern commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +286,109 @@ fn load_crash_kernel() -> Result<()> {
+    trace!("Serializied boot config string: {}", bootconfig);
+    let initrd = generate_initrd(&bootconfig.into_bytes())?;
+    trace!("Writing initrd image file: {:?}", initrd);
+    fs::write(BOOTCONFIG_INITRD_PATH, &initrd).context(error::WriteInitrdSnafu)?;
```
If it's async anyway, we could use tokio::fs

webern reviewed (approved):
Nice

arnaldo2792 commented on README.md:
````diff
@@ -549,6 +549,24 @@ Here are the metrics settings:
+init.param2 = ["hello", "goodbye"]
+```
+
+If an existing boot config exists at `/proc/bootconfig`, these settings' values' will be populated with values specified in the provisioned boot config file.
````
potential nit?: is there an extra `'` in values? or is that intended?

etungsten commented:
Push above and below addresses @webern 's comments.

etungsten commented:
Push above and below limits boot settings to bare metal variants only.

etungsten commented:
I need to figure out conditional compilation for certain parts of prairiedog that deals with boot settings that only exist for metal variants.

arnaldo2792 commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +286,119 @@ fn load_crash_kernel() -> Result<()> {
+where
+    P: AsRef<Path>,
+{
+    let boot_settings = schnauzer::get_settings(socket_path)
```
Is `schnauzer` the default client to get settings from the API? We don't use it across the binaries that interact with the API. Should we standardize it as the client going forward?

etungsten replied:
We're just borrowing `schnauzer`'s `get_settings` interface for getting API settings which itself is a thin wrapper around `apiclient::raw_request`. We don't use it across all the binaries since `schnauzer` was introduced later than the `apiclient` library and there are more specific use-cases that `schnauzer::get_settings` does not cover.

I don't think we need to enforce a "standard", there are other use cases where it's helpful to directly use the `apiclient` library call. In this case though, `schnauzer::get_settings` fits our use-case so we use it.

arnaldo2792 commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +286,119 @@ fn load_crash_kernel() -> Result<()> {
+
+fn generate_initrd(bootconfig: &[u8]) -> Result<Vec<u8>> {
+    let mut bootconfig_size = bootconfig.len();
+    // Calculate checksum of the boot config file
```
Nit: lesson learnt from my own PRs, explain the "why" instead of the "what", since the latter is obvious :+1:

arnaldo2792 commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +286,119 @@ fn load_crash_kernel() -> Result<()> {
+    Ok(initrd)
+}
+
+#[cfg(test)]
```
Nit: I'm not too much of a rustacean, but shouldn't the tests be grouped together, at the end of the file?

etungsten replied:
I don't think that has to be the case. You can have as many test modules as you'd like spread throughout the source file. I chose to keep the unit tests close to what the tests are actually testing.

arnaldo2792 commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +286,119 @@ fn load_crash_kernel() -> Result<()> {
+
+/// Reads `/proc/bootconfig` and populates the Bottlerocket boot settings based on the existing boot config data
+async fn generate_boot_settings() -> Result<()> {
+    let proc_bootconfig = match fs::read_to_string(PROC_BOOTCONFIG) {
```
A few lines above, you used `tokio::fs::<>`. Isn't that crate a good fit for this call as well? If not, when should we use `fs` vs `tokio::fs`?

arnaldo2792 commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +286,119 @@ fn load_crash_kernel() -> Result<()> {
+async fn generate_boot_settings() -> Result<()> {
+    let proc_bootconfig = match fs::read_to_string(PROC_BOOTCONFIG) {
+        Ok(s) => s,
+        Err(e) => {
```
Is it possible to use pattern matching in errors, like this?

```suggestion
        Err(io::ErrorKind::NotFound(e)) => Ok(()),
        Err(e) => { fail }
```

etungsten replied:
Unfortunately no. The matched type is of an `Result` which branch out to either `Ok` or `Err` but not the type of error since that's specific to the error type within the Error enum variant.

arnaldo2792 commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +286,119 @@ fn load_crash_kernel() -> Result<()> {
+        Ok(s) => s,
+        Err(e) => {
+            // If there's no `/proc/bootconfig`, then the user hasn't provisioned any kernel boot configuration.
+            if e.kind() == io::ErrorKind::NotFound {
```
Nice to have: you could refactor your function so that it receives `path`, and then you could unit test most of the ramifications.

etungsten replied:
I'm not sure what I would be unit testing here. Branching on the existence of a file? Or are you referring to `generate_boot_settings` as a whole? The deserialization has unit test in the `de` module. Or are you saying to unit test the json-serialized output?

arnaldo2792 replied:
Yeah, I meant the json-serialized output, but I consider it a nice-to-have :+1:

arnaldo2792 commented on sources/api/prairiedog/src/ser.rs:
```diff
@@ -0,0 +1,538 @@
+                "bark2".to_string() => vec!["woof".to_string()],
+                "silent".to_string() => vec!["".to_string()]
+            },
+            other: "burgermaster".to_string(),
```
:laughing: , nice

arnaldo2792 commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,127 @@ mod test_sysctl_key {
 
 // =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
+/// BootConfigKey represents a string that is a valid Kernel boot config keyword; Each key word must
```
Nit: lower case after `;`? (I might be wrong?)

arnaldo2792 commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,127 @@ mod test_sysctl_key {
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// BootConfigValue represents a string that is a valid Kernel boot config value; Each value only
```
Nit: Probably same as above?

etungsten commented:
Push above add conditional compilation for boot config related functionalities in prairedog.

zmrow commented on sources/models/src/lib.rs:
```diff
@@ -293,6 +294,25 @@ struct KernelSettings {
+        alias = "kernel",
+        rename(serialize = "kernel"),
+        default,
+        skip_serializing_if = "Option::is_none"
```
The struct member's value isn't an `Option` so I think this may be a typo?  Should it be `HashMap::is_empty`, or excluded all together?

etungsten replied:
With the `#[model]` attribute, all struct member fields are wrapped in `Option` by default. See https://github.com/bottlerocket-os/bottlerocket/blob/42bffd64bf9733394a317f961e6ff01852efe184/sources/models/model-derive/src/lib.rs#L152-L163

zmrow commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,127 @@ mod test_sysctl_key {
+    #[test]
+    fn valid_bootconfig_keyword() {
+        for ok in &[
+            "keyword",
```
Nit: I don't think the documentation explicitly disallows words that start with `-` or `_`, so let's add an extra test here to cover that.

zmrow commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,127 @@ mod test_sysctl_key {
+        let quoted = (input.starts_with('"') && input.ends_with('"'))
+            || (input.starts_with('\'') && input.ends_with('\''));
+        let chars_that_require_quotes = ['\'', '"', '\n', ',', ';', '#', '}'];
+        let valid_value = input
```
This is nice - I was wondering how this code was going to be to write - nice work!

zmrow commented on sources/api/prairiedog/src/bootconfig.rs:
```diff
@@ -0,0 +1,194 @@
+    info!("Generating initrd boot config from boot settings");
+    trace!("Boot settings: {:?}", boot_settings);
+    let bootconfig = to_string(&boot_settings).context(error::SerializeBootConfigSnafu)?;
+    trace!("Serializied boot config string: {}", bootconfig);
```
s/Serializied/Serializing/

zmrow commented on sources/api/prairiedog/src/bootconfig.rs:
```diff
@@ -0,0 +1,194 @@
+        + size_of::<u32>() * 2
+        + BOOTCONFIG_MAGIC.as_bytes().len();
+    let mut initrd = bootconfig.to_owned();
+    // initrd image file needs to be 4-byte aligned, so we add padding as necessary.
```
An extra comment or two here about the math would be helpful! :) (since we don't do bitwise stuff often)

zmrow commented on sources/api/prairiedog/src/lib.rs:
```diff
@@ -0,0 +1,9 @@
+//! The prairiedog library contains serializer and deserialzier implementations for kernel boot config files
+
+#![deny(rust_2018_idioms)]
+#[cfg(bottlerocket_platform = "metal")]
```
Hm - this works for the short term, but newer AWS/VMware variants may also have the ability to use this (since they'll be built with a newer grub)...

zmrow commented on README.md:
````diff
@@ -549,6 +549,26 @@ Here are the metrics settings:
+
+```toml
+[settings.boot]
+kernel.console = ["tty0", "ttyS1"]
````
Let's add a more complicated example including baud rate, flow control, etc.for the serial console:
`ttyS1,115200n8`

zmrow commented on README.md:
````diff
@@ -549,6 +549,26 @@ Here are the metrics settings:
     "vm.max_map_count" = "262144"
     ```
 
+#### Boot-related settings for bare metal variants
````
Once again, for the short term this will be only on metal, but longer term probably not.

We should talk a bit more about how is the best way to do this.

bcressey commented on sources/api/prairiedog/README.md:
```diff
@@ -2,11 +2,15 @@
   - loads the crash kernel from /boot
   - creates memory dumps when the kernel panics
+  - generates kernel boot config from settings
+  - generate settings from the existing kernel boot config file
```
couple nits:
```suggestion
It does the following:
  - _digs_ to find the active boot partition and mounts it in /boot
  - loads the crash kernel from /boot
  - creates memory dumps when the kernel panics
  - generates kernel boot config from settings
  - generates settings from the existing kernel boot config file
```

bcressey commented on sources/api/prairiedog/build.rs:
```diff
@@ -0,0 +1,54 @@
+use std::{env, process};
+
+fn main() {
+    // The code below emits `cfg` operators to conditionally compile this program based on the
+    // current variant.
+    // TODO: Replace this approach when the build system supports ideas like "variant
+    // tags": https://github.com/bottlerocket-os/bottlerocket/issues/1260
```
Rather than conditionally compile `prairiedog`, I think we can just build it with all the required functionality and just trust that it won't be invoked if the variant doesn't have the settings in its model.

etungsten replied:
Only metal variants right now have `boot: BootSettings` in its setting struct. In `prairiedog::generate_boot_config`, I'm fetching the boot settings I need to serialize the bootconfig.

So if I try building a variant that doesn't have `BootSettings` in its settings struct, then I run into this:
```
error[E0609]: no field `boot` on type `Settings`
  --> api/prairiedog/src/bootconfig.rs:57:25
   |
57 |         .and_then(|s| s.boot)
   |                         ^^^^ unknown field
   |
   = note: available fields are: `motd`, `kubernetes`, `updates`, `host_containers`, `bootstrap_containers` ... and 9 others
```

I could also add that field to all variant's settings struct. But that would mean it would be an inert setting that won't do anything for all existing variants (sans metal).

etungsten replied:
Ok, I think I can avoid this error by doing `apiclient::raw_request` and going through the JSON response body and check via `serde_json::Value` and see if `boot` exists under `settings` or not. That would avoid all the conditional compilation and we'd just output an error message for boot configuration subcommands in variants that don't support it.

bcressey commented on sources/models/src/modeled_types/mod.rs:
```diff
@@ -29,6 +29,18 @@ pub mod error {
+        InvalidBootconfigKeyword { input: String },
+
+        #[snafu(display(
+            "Kernel boot config values may only contain ASCII printable characters except for delimiters such as ';', '\n', ',', '#', and '}}'. Received '{}'",
```
```suggestion
            "Kernel boot config values may only contain ASCII printable characters except for delimiters such as ';', '\n', ',', '#', and '}'. Received '{}'",
```

etungsten replied:
Oh, I did `}}` because it needs to be escaped since it's a format string and `{` and `}` are reserved tokens.

bcressey commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,127 @@ mod test_sysctl_key {
+            "plain",
+            "@yogurt@",
+            "\"abc\"",
+            "\"
```
Should also check that the empty string is a valid value, since we'll need to be able to handle parameters like `quiet` that don't have a value.

bcressey commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,127 @@ mod test_sysctl_key {
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// BootConfigValue represents a string that is a valid Kernel boot config value; Each value only
+/// contains printable characters or spaces except for delimiters such as semicolon, newline, comma,
+/// hash, and closing brace. These delimiters are only usable if the value itself is quoted with
+/// single-quotes or double-quotes.
+/// BootConfigValue stores the original string and makes it accessible through standard traits.
```
I am not sure that these quoting rules are applied in the right place, or that the value shape is right.

For command line options like these:
```
slub_debug[=options[,slabs][;[options[,slabs]]...]
usbcore.quirks=0781:5580:bk,0a5c:5834:gij
```

We might input those as:
```
[settings.boot.kernel-parameters]
"slub_debug" = [
  "options,slabs",
  "options,slabs",
]
"usbcore.quirks" = [
  "0781:5580:bk",
  "0a5c:5834:gij"
]
```

(The separator between the two is different; `slub_debug` uses semicolons and `usbcore.quirks` uses commas. But that doesn't seem to matter for the purposes of transforming it into valid bootconfig.)

I think what we're saying here is if we had a setting like this:
```
"usbcore.quirks" = [
  ";}",
  "#,"
]
```

Then we'd have to write those bootconfig values as `";}"` and `'#,'` instead of `;}` and `#,`. That feels like a concern we should handle on the serialization side when writing out a value, rather than at this point where we're just concerned with getting tokens that _can_ be written out in a valid way.

On the key side, `usbcore.quirks` is problematic because it's an illegal keyword. I don't think we want to allow all of bootconfig's flexibility, but we do need to support at least one way of inputting data like this.

This seems to match our desired structure pretty well, as long as we allow dots in our keys:
```
foo.bar.baz = value1
foo.bar.qux.quux = value2
```

While this would seem to need a map of keys to any of the following:
* a value
* a list of values
* a nested map of keys
```
foo.bar {
   baz = value1
   qux.quux = value2
}
```

etungsten replied:
For reference, the settings model currently looks like this:
```
// Kernel boot settings
#[model]
struct BootSettings {
    #[serde(...)]
    kernel_parameters: HashMap<BootConfigKeyword, Vec<BootConfigValue>>,
    #[serde(...)]
    init_parameters: HashMap<BootConfigKeyword, Vec<BootConfigValue>>,
}
```

> While this would seem to need a map of keys to any of the following:
>   - a value
>   - a list of values
>   - a nested map of keys

For keys, I'll go ahead replace `BootConfigKeyword` with `BootConfigKey` and allow dotted keys as long as each individual keyword is still valid. I might need to change some things in the serialization module to accommodate, but it should be trivial. I think deserializing into `BootSettings` might be more complicated since the `Deserializer` would be expecting a nested struct/map when encountering dotted keys. It's not clear to me at the moment how to address that.

For values, I'll always assume the value being specified via settings is in its "quoted-form" and then do rudimentary checks on making sure it's ASCII. During serialization to the boot config format, all values will be quoted.

etungsten replied:
I was wondering if I could express the map value as a enum with enum variants mapping to a single boot config value, list of boot config values, or another map. But digging into our datastore serialization and deserialization implementation, we explicitly disallow and error upon tuple variants. So that wouldn't work.
https://github.com/bottlerocket-os/bottlerocket/blob/42bffd64bf9733394a317f961e6ff01852efe184/sources/api/datastore/src/serialization/pairs.rs#L200-L205

etungsten replied:
I think I have a path forward. I forgot I could just implement a custom function that implements a custom visitor for the Maps to deserialize into `BootSettings` fields like we do for k8s node taints and registry mirrors. Gonna try that out and see.

etungsten replied:
I ended up addressing this by adding a new parameter to the primary deserialization interface that specifies known prefixes for boot config keys. Now keys like `usbcore.quirks` are treated as a single key segment in the datastore.

etungsten commented:
Push above and below and tries to address comments.

Unit tests all pass.

etungsten commented:
Push above simplifies padding calculation a bit and addresses some leftover comments.

etungsten commented:
Push above rebases onto develop to pull the grub changes.

etungsten commented:
Push above removes some extraneous comments and fixes a log message that was causing sundog to choke.

etungsten committed:
sundog: update settings retrieval to use prefix

This updates the settings retrieval to use prefixes for matching and
querying. We cannot use full keys for matching because we now have
settings generator for keys that map to top level structs that
encompasses multiple settings.

etungsten commented:
Push above does the following: 
- adds the `api_exec_t` label to `thar-be-settings` so the process can transition to an selinux context that lets it spawn child processes, `prairiedog` in this case, that can write to the `BOTTLEROCKET-PRIVATE` partition. 
- adds a new commit to fix a `sundog` limitation when filtering settings when checking what needs to be generated.
- adds some clarification to the settings documentation.

Fully tested changes with `metal-dev` variant and it works as described.

etungsten commented:
Push above changes `std::fs::write` to `tokio::fs:write` which is just an async version of it. 

Tested and it still works as expected.

etungsten committed:
metal: remove kernel console settings from variant metadata

We've added bootconfig support so console settings should be specified
through there. The console settings specified to the kernel commandline
will always take effect, so we're removing them from the variant
metadata.

webern reviewed (approved):
nit: sources/api/prairiedog/src/ser.rs has inconsistent newline formatting between function definitions

zmrow commented on README.md:
```diff
@@ -555,6 +555,41 @@ Here are the metrics settings:
+
+*Please note that boot settings only exist for bare-metal variants at the moment*
+
+Specifying either of the following settings will generate a kernel boot config file to be loaded on subsequent boots:
```
Let's link to the kernel documentation here, I'm betting this feature isn't widely known.

Also - let's  expand on the expected format a bit and/or link to the docs on this as well.

zmrow commented on packages/selinux-policy/fs.cil:
```diff
@@ -45,6 +45,8 @@
 (filecon "/.*/usr/bin/storewolf" file api_exec)
 (filecon "/.*/usr/bin/cfsignal" file api_exec)
+(filecon "/.*/usr/bin/prairiedog" file api_exec)
+(filecon "/.*/usr/bin/thar-be-settings" file api_exec)
```
Paging @bcressey / @arnaldo2792  - this seem ok to you folks? :)

arnaldo2792 replied:
I talked about this with @etungsten offline and I believe this is the correct direction. Both binaries have to be `api_exec` so that they can enter the `api_t` domain, which has the permissions to write to the private partition.

bcressey replied:
`thar-be-settings` is ordinarily invoked by `apiserver`, which runs as `api_t`, but it's also invoked as a standalone binary in startup, where it runs as the default `system_t`. Transition rule makes sense - one of the settings it's applying can now trigger a write to the datastore filesystem.

`prairiedog` is invoked in two contexts:
* as a setting generator by sundog; it will have the `system_t` label, but it shouldn't need to write to the datastore filesystem here, since it's just adding settings via API
* as a restart command; it should inherit the `api_t` label since it's spawned by `apiserver`, which should allow it to write to the datastore filesystem.

So I'm not sure why `prairiedog` needs the label.

zmrow reviewed (commented):
Just a few small nits and I think we're there!

zmrow reviewed (approved):
:nerd_face:

arnaldo2792 commented on sources/api/prairiedog/src/de.rs:
```diff
@@ -0,0 +1,671 @@
+                } else {
+                    quote = Some(c);
+                }
+            } else if c == ',' && quote == None {
```
Nit: `s|quote == None|quote.is_none()|`

arnaldo2792 commented on sources/api/prairiedog/src/de.rs:
```diff
@@ -0,0 +1,671 @@
+            }
+        );
+
+        // Make sure unbalanced quotes fails in sequences
```
Nit: these tests could be in their own functions with names that suggest what the test is actually checking, ex. `fails_with_unbalanced_quotes`, `fails_without_comma_delimiters`

arnaldo2792 commented on sources/models/src/modeled_types/mod.rs:
```diff
@@ -29,6 +29,18 @@ pub mod error {
         InvalidIdentifier { input: String },
 
+        #[snafu(display(
+            "Kernel boot config keywords may only contain ASCII alphanumerics plus hyphens and underscores, received '{}'",
```
Nit: there is a similar text a few lines above, I think you could be consistent and use (almost?) the same message.

etungsten replied:
I'd rather keep the error variants separate since they're for different model types that have similar but different format requirements.

bcressey commented on README.md:
````diff
@@ -555,6 +555,41 @@ Here are the metrics settings:
+*Please note that boot settings only exist for bare-metal variants at the moment*
+
+Specifying either of the following settings will generate a kernel boot config file to be loaded on subsequent boots:
+* `settings.boot.kernel-parameters`: These allows additional kernel parameters to be specified to the kernel command line during boot.
+* `settings.boot.init-parameters`: These allows additional init parameters to be specified to init during boot.
+
+Example user data for specifying boot settings:
+
+```toml
+[settings.boot.kernel]
````
The documentation and examples are inconsistent - is it `kernel` or `kernel-parameters`?

etungsten replied:
Both are acceptable since I added `kernel` as an alias for `kernel-parameters` to serde. You're right that it should probably be consistent in the documentation.

bcressey commented on sources/api/sundog/src/main.rs:
```diff
@@ -245,8 +244,18 @@ where
+
         // Don't clobber settings that are already populated
-        if populated_settings.contains(&setting) {
+        // We're checking for prefix matches here because some generators are for settings that map
+        // to a top level struct that might have subfields. For example, a settings generator for
+        // `settings.boot` encompasses both `settings.boot.kernel` and `settings.boot.init`.
```
To check my understanding -

We could use `prairiedog` as a generator for `settings.boot.kernel-parameters` and `settings.boot.init-parameters`, but that would result in two program invocations rather than just one, and we'd essentially have to do the same work twice, throwing away half of it each time.

If we had functionality like restart commands have, where we could pass in a list of settings to generate, then that might work, but we don't. And where restart commands initially operated in a batch mode ("do everything") and we later added a more targeted mode for host containers ("do just this part"), settings generators have always operated on a single setting and it'd be a pretty big shift to try to coalesce calls to the same generator.

etungsten replied:
> We could use `prairiedog` as a generator for `settings.boot.kernel-parameters` and `settings.boot.init-parameters`, but that would result in two program invocations rather than just one, and we'd essentially have to do the same work twice, throwing away half of it each time.

If we're talking about before this change, that would be correct. We'd have to specify `metadata.settings.boot.kernel-parameters.setting-generator` and `metadata.settings.boot.init-parameters.setting-generator` separately.

But with the prefix matching, we can specify settings generator for a top-level setting like `settings.boot` and `prairiedog` would just be invoked once to return a serialized json string that contains everything under `settings.boot`. Another neat thing is if either `kernel-parameters` or `init-parameters` is already set, then sundog would skip over the settings generation for the top-level setting, i.e. `settings.boot`.

bcressey commented on sources/models/shared-defaults/boot.toml:
```diff
@@ -0,0 +1,9 @@
+# Boot related settings
+
+[metadata.settings.boot]
+affected-services = ["bootconfig"]
+setting-generator = "/usr/bin/prairiedog generate-boot-settings"
+
+[services.bootconfig]
+configuration-files = []
+restart-commands = ["/usr/bin/prairiedog generate-boot-config"]
```
Would this setting generator still be invoked if we have `settings.boot.kernel-parameters` and `settings.boot.init-parameters` defined?

(Maybe that's handled correctly.)

etungsten replied:
If either of `settings.boot.kernel-parameters` or `settings.boot.init-parameters` is defined, sundog would skip over settings generation for `settings.boot` to avoid clobbering it by checking if any populated_setting has the target setting as prefix, in this case, `settings.boot`. Since the generation would be for everything under `settings.boot` and prairiedog only returns the whole `settings.boot` serialized json output.

bcressey commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,128 @@ mod test_sysctl_key {
+            keyword
+                .chars()
+                .all(|c| (c.is_ascii() && c.is_alphanumeric()) || c == '-' || c == '_')
+                && !keyword.is_empty()
```
nit: might be slightly more efficient to reverse these checks, assuming the test for an empty string is cheap.

bcressey commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,128 @@ mod test_sysctl_key {
+    fn try_from(input: &str) -> Result<Self, error::Error> {
+        ensure!(
+            input.chars().all(|c| c.is_ascii()),
+            error::InvalidBootconfigValueSnafu { input }
```
This seems like it would allow the control characters as well, which I think we want to exclude.

`0x0A` (newline) seems like it's meant to be allowed in values, but I don't see a use case.

bcressey commented on sources/models/src/modeled_types/mod.rs:
```diff
@@ -29,6 +29,18 @@ pub mod error {
+        InvalidBootconfigKey { input: String },
+
+        #[snafu(display(
+            "Kernel boot config values may only contain ASCII printable characters except for delimiters such as ';', '\n', ',', '#', and '}}'. Received '{}'",
```
This message isn't correct now, right? We do allow these delimiters.

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -46,12 +55,23 @@ const KEXEC_CMD_LINE: &str = "maxcpus=1 systemd.unit='capture-kernel-dump.servic
+const BOOTCONFIG_MAGIC: &str = "#BOOTCONFIG\n";
+// 1 << 2 = 4
+const BOOTCONFIG_ALIGN: usize = 0b100;
+const BOOTCONFIG_INITRD_PATH: &str = "/var/lib/bottlerocket/bootconfig.data";
```
nit: it might be worth putting the initrd builder code in a different module, since it's somewhat likely to grow in scope over time.

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +285,130 @@ fn load_crash_kernel() -> Result<()> {
+    P: AsRef<Path>,
+{
+    // Only generate the initrd if the boot settings exist
+    if let Some(boot_settings) = get_boot_config_settings(socket_path).await? {
```
Can we generate an empty bootconfig if we don't have boot settings? If we ever make it possible to delete settings, we'll want to be able to reset it back to the empty file we create at build time.

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -263,37 +417,42 @@ fn setup_logger(args: &Args) -> Result<()> {
 /// ＜コ：ミ くコ:彡 ＜コ：ミ くコ:彡 ＜コ：ミ くコ:彡 ＜コ：ミ くコ:彡 ＜コ：ミ くコ:彡 ＜コ：ミ くコ:彡
 mod error {
-    use nix;
+    use prairiedog::{de, ser};
```
nit: if we're just doing this so we can write `ser::Error`, I think the qualified type name `prairiedog::ser::Error` would be more clear.

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -12,12 +16,17 @@
+use bytes::BufMut;
+use model::BootSettings;
+use prairiedog::de::from_str;
+use prairiedog::ser::to_string;
```
I'd definitely prefer the fully qualified names for these; they're too generic otherwise.

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +285,130 @@ fn load_crash_kernel() -> Result<()> {
+    if let Some(boot_settings) = get_boot_config_settings(socket_path).await? {
+        info!("Generating initrd boot config from boot settings");
+        trace!("Boot settings: {:?}", boot_settings);
+        let bootconfig = to_string(&boot_settings).context(error::SerializeBootConfigSnafu)?;
```
It seems like the convention in `serde_json` and `toml` is to return a `Value` struct that has a `to_string()` method. Could also implement `to_vec()` to get the byte array back.

It might feel more polished to have a `bootconfig` or `serde-bootconfig` crate that tried to mimic the conventional interfaces.

etungsten replied:
I'm not sure if it's convention to ser/de to `Value` types. It's provided as a way to ser/de to/from "untyped" recursive enum representations in the absence of strongly-typed representations. In both `serde_json` and `toml-rs`, you're able to ser/de directly to those strongly-typed structures like what we're doing here with bootconfig, e.g., https://docs.serde.rs/serde_json/#parsing-json-as-strongly-typed-data-structures

I agree that there is a convenience in being able to ser/de into `Value` types that can then be traversed freely, but that involves revamping the ser/de implementation here and I don't see a current need for that. Is it ok if we track this as an issue to come back to?

Regarding moving the ser and de modules out to a completely separate crate from prairiedog, I can do that if you'd like in this PR.

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +285,130 @@ fn load_crash_kernel() -> Result<()> {
+    let settings: serde_json::Value =
+        schnauzer::get_json(socket_path, uri, None as Option<(String, String)>)
+            .await
+            .context(error::RetrieveSettingsSnafu)?;
```
Can we pass a query parameter to `schnauzer::get_json` to avoid getting the entire serialized structure back when we only need part of it?

(It would be nice for `schnauzer` to have a generic helper method that allowed us to peel off the outer structure and convert directly to the desired `T` via serde. Not a blocker here but if you saw a quick way to implement that, it'd be nice!)

bcressey reviewed (commented):
I still need to chew on the serde implementation, but reviewed the rest of it.

etungsten commented:
Push above rebases onto develop to fix conflicts

etungsten committed:
prairiedog: add boot config functionality

 - Adds new subcommand `generate-boot-config` to generate an initrd
image file with boot config contents populated from Bottlerocket boot
settings.
 - Adds new subcommand `generate-boot-settings` to generate bottlerocket
settings from kernel's bootconfig file at `/proc/bootconfig`

Moves prairiedog from `sources/prairiedog` to `sources/api/prairiedog`
now that it needs to query from the Bottlerocket API and need settings
model information.

etungsten commented:
Push above adds quite a lot of changes:

- Addresses @zmrow 's comments about adding a link to the boot config docs in README.md
- Removes the `api_exec` label from prairiedog since it doesn't need it and can inherit it from `thar-be-settings` when it needs to write to the private partition.
- Separates out Bootconfig serde implementation into its own separate `bootconfig` crate
- Create a separate module within prairiedog for initrd specific logic
- Adds more unit tests targetting JSON serialized settings during settings generation
- Adds more unit tests to cover more cases
- Adds prefix as a query param when looking up boot settings for generating bootconfig file
- Other miscellaneous dress-ups to address comments. 


Tested changes and they work as expected

etungsten commented:
Push above fixes builds for variants other than metal-*

cbgbt commented on sources/bootconfig/src/de.rs:
```diff
@@ -0,0 +1,690 @@
+        // Ensure the whole key is a valid boot config key
+        for key in &keys {
+            let key = key.join(".");
+            let valid_key = key.split('.').all(|keyword| {
+                !keyword.is_empty()
+                    && keyword
+                        .chars()
+                        .all(|c| (c.is_ascii() && c.is_alphanumeric()) || c == '-' || c == '_')
```
Does it make sense to put this check somewhere that can be shared with the model code that does the same check?

etungsten replied:
I feel like the snippet isn't that long and there really isn't a suitable place for it. I could add it as a public function in the `bootconfig` crate, then `model` would have to take a dependency on `bootconfig` just for that piece of logic. What do you think?

cbgbt replied:
I suppose the set of allowable characters is really stable. This is probably fine.

cbgbt commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,126 @@ mod test_sysctl_key {
+        }
+    }
+
+    #[test]
+    fn invalid_bootconfig_key() {
+        for err in &[
```
Does it make sense to check unicode codepoints here?

etungsten replied:
I think it does. Bootconfig keywords explicitly only allow ASCII alphabets. We have `①` in there as a test case, but I can add more if that's what's being asked.

cbgbt replied:
Oh, I missed that. No, this is fine.

cbgbt commented on sources/bootconfig/src/de.rs:
```diff
@@ -0,0 +1,690 @@
+init.silent1 = ""
+init.silent2 =
+kernel.corn1 = "a", "unbalanced, quotes
+kernel.corn2 = "hello""#,
```
Should't both of these fails? Should we separate them into different assertions?

cbgbt commented on sources/bootconfig/src/ser.rs:
```diff
@@ -0,0 +1,543 @@
+
+    fn serialize_bool(self, _v: bool) -> Result<()> {
+        unsupported_type("bool")
+    }
```
Unfortunate that there aren't macros for this repetition :frowning_face:

cbgbt commented on sources/bootconfig/src/ser.rs:
```diff
@@ -0,0 +1,543 @@
+    T: ?Sized + Serialize,
+{
+    // Preallocate string buffer
+    let mut output = String::with_capacity(128);
```
Curious how you chose the starting capacity. I think it makes sense, but I went down the rabbit hole of trying to find docs on how Rust chooses a default capacity and couldn't find what I was looking for!

etungsten replied:
It's mostly arbitrary. I took it from serde-json and toml-rs (e.g. [toml-rs](https://github.com/alexcrichton/toml-rs/blob/70caf4085324bff452d8c20ead55974513030696/src/ser.rs#L91)).  It's hard to come up with a heuristic for guesstimating how much we should preallocate.

Having a preallocated buffer avoids a lot of memory allocation calls since we'll be continuously appending to the output `String`.

cbgbt replied:
As a followup: I spent some more time peeking at the implementation of `Vec` and I believe by default `capacity` is set to 0 to delay allocations in cases where none occur.

Anyways, I totally buy in to the idea of having something preallocated here.

cbgbt reviewed (commented):
I've mostly spent time reviewing the custom ser/de stuff. I don't have any major problems, just a few testing nits and questions.

etungsten commented:
Push above revises some unit tests in the `bootconfig` crate to address @cbgbt 's comments.

zmrow commented on sources/bootconfig/README.md:
```diff
@@ -0,0 +1,15 @@
+
+## Background
+
+The bootconfig library contains serializer and deserialzier implementations for the kernel boot configuration format.
```
Nit: s/deserialzier/deserializer

zmrow commented on sources/bootconfig/README.md:
```diff
@@ -0,0 +1,15 @@
+
+The bootconfig library contains serializer and deserialzier implementations for the kernel boot configuration format.
+
+The serde implementation currently supports the boot config format in `/proc/bootconfig` as show [here](https://www.kernel.org/doc/html/latest/admin-guide/bootconfig.html#proc-bootconfig).
```
nit: s/as show/as shown

zmrow reviewed (approved):
Just a few nits!

bcressey commented on README.md:
```diff
@@ -560,6 +560,43 @@ Here are the metrics settings:
+
+Specifying either of the following settings will generate a kernel boot config file to be loaded on subsequent boots:
+* `settings.boot.kernel-parameters`: These allows additional kernel parameters to be specified to the kernel command line during boot.
+* `settings.boot.init-parameters`: These allows additional init parameters to be specified to init during boot.
```
nit:
```suggestion
* `settings.boot.kernel-parameters`: This allows additional kernel parameters to be specified on the kernel command line during boot.
* `settings.boot.init-parameters`: This allows additional init parameters to be specified on the kernel command line during boot.
```

bcressey commented on README.md:
````diff
@@ -560,6 +560,43 @@ Here are the metrics settings:
+"splash" = []
+```
+
+If an existing boot config exists at `/proc/bootconfig`, these settings' values will be populated with values specified in the provisioned boot config.
````
nit:
```suggestion
If boot config data exists at `/proc/bootconfig`, it will be used to generate these API settings on first boot.
```

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -46,12 +55,22 @@ const KEXEC_CMD_LINE: &str = "maxcpus=1 systemd.unit='capture-kernel-dump.servic
+const PROC_BOOTCONFIG: &str = "/proc/bootconfig";
+const DEFAULT_BOOTCONFIG_STR: &str = r#"kernel = ""
+init = ""
+"#;
```
nit: maybe this for readability?
```suggestion
const DEFAULT_BOOTCONFIG_STR: &str = r#"
kernel = ""
init = ""
"#;
```

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +284,126 @@ fn load_crash_kernel() -> Result<()> {
+        Some(boot_settings) => {
+            info!("Generating initrd boot config from boot settings");
+            trace!("Boot settings: {:?}", boot_settings);
+            let bootconfig = bootconfig::ser::to_string(&boot_settings)
```
Consider re-exporting some functions from `ser` and `de` as a convenience, like the `toml` crate does:
```suggestion
            let bootconfig = bootconfig::to_string(&boot_settings)
```

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +284,126 @@ fn load_crash_kernel() -> Result<()> {
+        None => {
+            // If we don't have any boot settings, write out an initrd with default boot config contents
+            trace!("Serializing boot config string: {}", DEFAULT_BOOTCONFIG_STR);
+            generate_initrd(&DEFAULT_BOOTCONFIG_STR.to_string().into_bytes())?
```
nit: could potentially de-duplicate this line if we get bytes rather than an initrd from the `let` assignment

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +284,126 @@ fn load_crash_kernel() -> Result<()> {
+fn boot_config_to_boot_settings_json(bootconfig_str: &str) -> Result<String> {
+    // We'll only send the setting if the existing boot config file fits our settings model
+    let boot_settings =
+        bootconfig::de::from_str::<BootSettings>(bootconfig_str, Some(&["kernel", "init"]))
```
Same re-export suggestion here -
```suggestion
        bootconfig::from_str::<BootSettings>(bootconfig_str, Some(&["kernel", "init"]))
```

(Although it doesn't seem idiomatic to take more than just a `str` as input to a `from_str` function.)

bcressey replied:
The more I think about it, the more unnatural it feels to deserialize directly into the `BootSettings` structure like this. That ends up polluting our deserializer with knowledge about the destination shape, which breaks the abstraction this interface is intended to provide.

The trouble is that the maps in the bootconfig file might reach an arbitrary depth; we could end up with:
* `Map<String, String>`
* `Map<Map<String, String>`
* `Map<Map<Map<String, String>`

And so on.

If we had a `Value` type then we could make `BootSettings` handle this, [maybe](https://gist.github.com/bcressey/496abca53539510f5f073f5d8182424e).

But you didn't want a `Value` type and I agree that's pretty complex. I'm just not sure if there's a middle ground. If we can't round-trip somewhat arbitrary data structures to and from the bootconfig format then I'm not confident `serde` is the right interface.

To take a step back - what if we had just two functions and no serde?

First function could turn a `String` read from `/proc/bootconfig` into a `BootSettings` struct, checking each line for either an `init` or `kernel` prefix, stripping that off, splitting the rest of the line on the first " = ", and putting the entry into the corresponding `HashMap<BootConfigKey, Vec<BootConfigValue>>`. We'd use this during settings generation.

Second function could turn a `BootSettings' struct into a `String` in the obvious way. We'd use this during initrd generation.

Do we need more flexibility than that?

etungsten replied:
I agree the interface here isn't the best.

I think that's reasonable. The original thinking was that we'd implement this with serde with the goal to eventually expand it so it can be used elsewhere, e.g. an CLI tool for generating/validating bootconfig files. Maybe that's a case of YAGNI and we can avoid all that complexity and code for now. I'm also not too happy even with the complexity with what I have right now and maybe I should have thought more critically about simplifying this.

I'll go ahead with direct parsing so this PR should slim down a lot more.

bcressey replied:
No worries! I also started out feeling like serde made the most sense. I think the moment it stopped being a natural fit was when we realized that the `BootSettings` struct couldn't just be a pure deserialization because of the arbitrary nesting depth of kernel parameters - but at that point the serde implementation was already done and hard to un-see.

I've certainly learned a lot more about serde from studying this PR, so I consider it a valuable experience and not a mistake. But I agree that YAGNI is the right call.

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -253,8 +284,126 @@ fn load_crash_kernel() -> Result<()> {
+    static BOOTCONFIG_STR: &str = r#"kernel.console = "ttyS1,115200n8", "tty0"
+init.systemd.log_level = "debug"
+init.splash = ""
+"#;
```
nit:
```suggestion
    static BOOTCONFIG_STR: &str = r#"
kernel.console = "ttyS1,115200n8", "tty0"
init.systemd.log_level = "debug"
init.splash = ""
"#;
```

(IIRC Rust trims leading whitespace so we could also indent the kernel, initrd lines.)

bcressey commented on sources/bootconfig/src/error.rs:
```diff
@@ -0,0 +1,74 @@
+    BareString,
+
+    #[snafu(display("Expected K-V compound types, encountered bare sequence"))]
+    BareSeq,
```
These could potentially be collapsed into a single variant, like:
```
#[snafu(display("Expected '{}', found '{}'", expected, found))]
    UnexpectedType {
        expected: String,
        found: String,
    },
```

bcressey commented on sources/bootconfig/src/error.rs:
```diff
@@ -0,0 +1,74 @@
+    #[snafu(display("Invalid boot config file, expected key-value entries for each line"))]
+    InvalidBootConfig,
+
+    #[snafu(display("Invalid boot config key encountered '{}'. Boot config keywords may only contain ASCII alphas, numbers, dashes or underscores", key))]
```
nit:
```suggestion
    #[snafu(display("Invalid boot config key '{}'. Boot config keywords may only contain ASCII alphas, numbers, dashes or underscores", key))]
```

bcressey commented on sources/bootconfig/src/error.rs:
```diff
@@ -0,0 +1,74 @@
+    InvalidKey { key: String },
+
+    #[snafu(display(
+        "Invalid boot config value encountered '{}'. Boot config values may only contain ASCII printable characters except for delimiters such as ';', '\n', ',', '#', and '}}'",
```
nit:
```suggestion
        "Invalid boot config value '{}'. Boot config values may only contain ASCII printable characters except for delimiters such as ';', '\n', ',', '#', and '}}'",
```

bcressey commented on sources/bootconfig/src/ser.rs:
```diff
@@ -0,0 +1,544 @@
+    }
+    fn serialize_i16(self, _v: i16) -> Result<()> {
+        unsupported_type("i16")
+    }
```
nit: I might still prefer newlines between these functions - surprised that `cargo fmt` doesn't add them. Might also be good to put them all in one spot, toward the end.

bcressey commented on sources/bootconfig/src/ser.rs:
```diff
@@ -0,0 +1,544 @@
+        }
+        self.output.push('\n');
+        Ok(())
+    }
```
This felt somewhat brittle to me. Because of the way we're quoting everything, we know that any value would end in `"` rather than `=`, so the `ends_with()` test is probably right - but that's specific to the current implementation rather than required by the format.

It might be a bit cleaner to maintain a counter of elements, setting it to zero when we start, incrementing it along the way, and zeroing it back out at the end. Then we could test the sequence count instead of the last character.

bcressey commented on sources/bootconfig/src/ser.rs:
```diff
@@ -0,0 +1,544 @@
+    where
+        T: Serialize,
+    {
+        let value_str = value.serialize(StringExtractor).unwrap();
```
It's not obvious to me why this `unwrap()` is guaranteed not to panic.

bcressey commented on sources/bootconfig/src/ser.rs:
```diff
@@ -0,0 +1,544 @@
+    {
+        trace!("Serializing map key at prefix {:?}", self.prefix);
+        // We're given a serializable thing; need to serialize it to get a string we can work with.
+        let key_str = key.serialize(StringExtractor).unwrap();
```
Same note here - are we sure this `unwrap()` cannot panic?

bcressey commented on sources/bootconfig/src/ser.rs:
```diff
@@ -0,0 +1,544 @@
+}
+
+#[cfg(test)]
+mod test {
```
Would [serde_test](https://serde.rs/unit-testing.html) be useful here?

bcressey commented on sources/bootconfig/src/ser.rs:
```diff
@@ -0,0 +1,544 @@
+
+/// Used to implement a serializer for components of the kernel boot config that must be a string.
+/// For example, the key words and the elements of a sequence.
+struct StringExtractor;
```
Consider using `serde_plain::to_string(value)` on the two "must be strings" cases instead - it seems to do almost exactly this, while also allowing numbers and booleans.

bcressey commented on sources/bootconfig/src/ser.rs:
```diff
@@ -0,0 +1,544 @@
+            }
+        } else {
+            self.key_words = Some(vec![key_str]);
+        }
```
Seems like we're missing some sort of check that the keys we're generating are valid bootconfig keys. Same with values - we're quoting the strings but there are characters we can't serialize, like `"`.

bcressey commented on sources/bootconfig/src/ser.rs:
```diff
@@ -0,0 +1,544 @@
+    }
+    fn serialize_i64(self, _v: i64) -> Result<()> {
+        unsupported_type("i64")
+    }
```
We could potentially consolidate these "unsupported type" functions using the same sort of macro that `serde_plain` uses to implement them.

bcressey commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,126 @@ mod test_sysctl_key {
+impl TryFrom<&str> for BootConfigKey {
+    type Error = error::Error;
+
+    fn try_from(input: &str) -> Result<Self, error::Error> {
+        // Each individual keyword must be valid
+        let valid_key = input.split('.').all(|keyword| {
+            !keyword.is_empty()
+                && keyword
+                    .chars()
+                    .all(|c| (c.is_ascii() && c.is_alphanumeric()) || c == '-' || c == '_')
+        });
+        ensure!(valid_key, error::InvalidBootconfigKeySnafu { input });
```
Assuming we validate in the bootconfig module, it might be nice to treat this as we do with base64, where we just go ahead and pass it to the bootconfig routines and see if we get an error back.

Something like:
```
bootconfig::from_str(format!("{} = \"\", input)?;
Ok(BootConfigKey {
    inner: input.to_string(),
})
```

(Same for values.)

bcressey commented on sources/bootconfig/src/de.rs:
```diff
@@ -0,0 +1,683 @@
+//!
+//! Note that a lot of the logic for deserializing a nested structure is borrowed from
+//! Bottlerocket's datastore deserializer since we're essentially solving the same problem but
+//! there are enough differences to not directly depend on the datastore's implementation of this.
```
nit: I'm not sure this comment will age very well - might be better to remove it and just treat them as two distinct bodies of code now.

etungsten commented:
Push above addresses @bcressey comments.

- Removes the bootconfig serde implementation, the `bootconfig` crate is now gone.
- Now parses and serializes bootconfig files in prairiedog directly to/from `models::BootSettings` without serde.

Tested all the changes and they still work as expected:
```
[erikson@admin]$ apiclient get settings.boot
{
  "settings": {
    "boot": {
      "init": {
        "systemd.log_level": [
          "debug"
        ]
      },
      "kernel": {
        "console": [
          "tty0",
          "ttyS1,115200n8"
        ]
      }
    }
  }
}
```

```
[erikson@admin]$ apiclient set --json '{"boot":{"kernel":{"console":["tty0","ttyS1,115200n8"]},"init":{"systemd.log_level":["info"]}}}'
[erikson@admin]$ sudo sheltie
bash-5.1# cat /var/lib/bottlerocket/bootconfig.data 
kernel.console = "tty0", "ttyS1,115200n8"
init.systemd.log_level = "info"
L#BOOTCONFIG
```

etungsten commented:
Push above does some minor refactoring and clean-up

etungsten commented:
Push above fixes initrd generation that I missed in my initial testing.

bcressey commented on sources/api/prairiedog/src/main.rs:
```diff
@@ -1,26 +1,34 @@
 use std::path::Path;
 use std::process::{self, Command};
 
+mod bootconfig;
+mod error;
+mod initrd;
+
```
Do we still need these?

etungsten replied:
Yes we do, otherwise they don't get included in prairedog's module tree.

bcressey commented on sources/api/prairiedog/src/bootconfig.rs:
```diff
@@ -0,0 +1,410 @@
+        if i > 0 {
+            output.push_str(", ");
+        }
+        output.push_str(&format!("\"{}\"", v));
```
If the value has a double-quote in it, I think we need to wrap it in single quotes instead.

bcressey commented on sources/api/prairiedog/src/bootconfig.rs:
```diff
@@ -0,0 +1,410 @@
+    let mut start_index = 0;
+    for (i, c) in input.trim().chars().enumerate() {
+        if expect_delimiter && !c.is_whitespace() && c != ',' {
+            // return error::ExpectedArrayCommaSnafu { input: values }.fail();
```
This is commented out?

bcressey commented on sources/api/prairiedog/src/bootconfig.rs:
```diff
@@ -0,0 +1,410 @@
+                    .context(error::ParseBootConfigKeySnafu)?,
+                values,
+            );
+        } else if !key.starts_with("kernel") && !key.starts_with("init") {
+            return error::UnsupportedBootConfigKeySnafu { key }.fail();
+        }
```
Should we always error in this case?
```suggestion
        } else {
            return error::UnsupportedBootConfigKeySnafu { key }.fail();
        }
```

Otherwise I think we'd allow input like:
```
kernel = foo
kernel.foo = bar
init = bar
init.bar = baz
```

etungsten replied:
> kernel = foo
kernel.foo = bar
init = bar
init.bar = baz

I think we want to allow that. That's a valid boot config, it's just that later entries will overwrite the previous one.

I need that special case there for our default bootconfig which is just:
```
kernel = ""
init = ""
```

etungsten replied:
Slight correction, it's not overwriting, but 
> sub-keys and a value can coexist under a parent key. For example, following config is allowed.:
foo = value1
foo.bar = value2

from https://www.kernel.org/doc/html/latest/admin-guide/bootconfig.html#same-key-values

etungsten replied:
What I really should do here is special case allow the default bootconfig, but disallow `kernel` and `init` as parents keys for any non-null value, since `BootSettings` can't accommodate that. Currently, if there's a non-null value, it's just completely ignored.

So allow `kernel = ""` but disallow `kernel = foo`

bcressey commented on sources/api/prairiedog/src/bootconfig.rs:
```diff
@@ -0,0 +1,410 @@
+            init_parameters: None,
+        };
+        assert_eq!(
+            serialize_boot_settings_to_boot_config(&boot_settings).unwrap(),
+            r#""#
```
Just double-checking - the kernel is fine with an empty bootconfig file? Or do we need to emit our placeholder value to avoid a warning?

The `bootconfig` tool in the kernel tree says "Error: Config data is empty" if given an empty file.

etungsten replied:
`prairiedog` handles the case where BootSettings is None or if both `kernel` and `init` are None.
It stubs in the default bootconfig file if the settings aren't there:
```
bash-5.1# apiclient get boot
{}
bash-5.1# prairiedog --log-level trace generate-boot-config
22:34:02 [TRACE] (1) schnauzer: [api/schnauzer/src/lib.rs:81] GETing from /settings?prefix=boot
...
22:34:02 [TRACE] (1) schnauzer: [api/schnauzer/src/lib.rs:95] JSON response: {}
22:34:02 [TRACE] (1) prairiedog::bootconfig: [api/prairiedog/src/bootconfig.rs:67] Serializing boot config string: 
    kernel = ""
    init = ""
...
```

There are no case where we would ever serialize out an empty bootconfig file

bcressey commented on sources/api/prairiedog/src/bootconfig.rs:
```diff
@@ -0,0 +1,410 @@
+        let output = lines.join("\n");
+        assert_eq!(
+            output,
+            r#"init.splash = ""
+init.systemd.log_level = "debug"
+kernel.console = "ttyS1,115200n8", "tty0""#
```
nit: all of these tests would be more readable if given the same initial-newline and indent treatment as `STANDARD_BOOTCONFIG_STR` below.

etungsten replied:
I'd have to use something like [this crate](https://docs.rs/unindent/latest/unindent/#) to ensure the leading whitespace is removed from the raw string.


`STANDARD_BOOTCONFIG_STR`'s situation is different because I `trim()` each line before parsing them down to K-Vs.

etungsten replied:
I tried making this better with some extra `trim()`ing.

bcressey commented on sources/api/prairiedog/src/bootconfig.rs:
```diff
@@ -0,0 +1,410 @@
+    }
+
+    static UNSUPPORTED_BOOTCONFIG_STR: &str = r#"
+        do.andriods.dream.of.electric.sheep = "?"
```
nit:
```suggestion
        do.androids.dream.of.electric.sheep = "?"
```

bcressey commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -647,6 +647,126 @@ mod test_sysctl_key {
+
+    #[test]
+    fn invalid_bootconfig_value() {
+        for err in &["bottlerocket①", "💝", "Ï", "—"] {
```
I think a value with both a single quote and a double quote should also be illegal, since quotes can't be escaped.

bcressey commented on sources/api/prairiedog/src/bootconfig.rs:
```diff
@@ -0,0 +1,410 @@
+    let chars_that_require_quotes = ['\'', '"', '\n', ',', ';', '#', '}'];
+    let valid_value = input
+        .chars()
+        .all(|c| c.is_ascii() && (quoted || !chars_that_require_quotes.contains(&c)));
```
It might be better to get rid of these checks entirely.

We can remove the outer quotes (if present) and then delegate the work to the `try_into()` call at the end, which will repeat these tests.

That would also let us remove the `InvalidBootConfigValueSnafu` error here which has different criteria than the modeled type.

etungsten replied:
I think we need these checks to guard against unquoted values with disallowed delimiters. 
>Each value only contains printable characters or spaces except for delimiters such as semi-colon (;), new-line (\n), comma (,), hash (#) and closing brace (}).

The modeled type does not check for that, it assumes all `BootConfigValues` are eventually double/single quoted so it only ensures the value is printable ASCII sans control chars.

For example, if we're parsing a boot config file:
```
kernel = #bang
```

Without this check, it'll be allowed.

bcressey replied:
Ah, gotcha. Even though we could accept that value - since we know we'll end up quoting it - it's not a valid input so we should reject it.

bcressey reviewed (commented):
Looks good to me except for some fussing about single- and double quotes.

etungsten commented:
Pushes above addresses @bcressey 's comments.

- `prairiedog` will now error when parsing boot config files that have `kernel` and `init` as parent keys to a list of non-null values. This is disallowed since `BootSettings` model does not support it.
- `BootConfigValue` now rejects value strings that mix and contain both single quotes and double quotes. Quotes cannot be escaped in boot config values.

bcressey reviewed (approved):
Nice!

👢 🔧

zmrow reviewed (approved):
:octopus: 

Nice work!
